bindings/rust/s2n-tls/src/cert_chain.rs

```diff
@@ -6,34 +6,122 @@ use s2n_tls_sys::*;
 use std::{
     marker::PhantomData,
     ptr::{self, NonNull},
+    sync::Arc,
 };
 
+/// Internal wrapper type used for a convenient drop implementation.
+///
+/// [CertificateChain] is internally reference counted. The reference counted `T`
+/// must have a drop implementation.
+struct CertificateChainHandle {
+    cert: NonNull<s2n_cert_chain_and_key>,
+    is_owned: bool,
+}
+
+impl CertificateChainHandle {
+    fn from_owned(cert: NonNull<s2n_cert_chain_and_key>) -> Self {
+        Self {
+            cert,
+            is_owned: true,
+        }
+    }
+
+    fn from_reference(cert: NonNull<s2n_cert_chain_and_key>) -> Self {
+        Self {
+            cert,
+            is_owned: false,
+        }
+    }
+}
+
+impl Drop for CertificateChainHandle {
+    fn drop(&mut self) {
+        // ignore failures since there's not much we can do about it
+        if self.is_owned {
+            unsafe {
+                let _ = s2n_cert_chain_and_key_free(self.cert.as_ptr()).into_result();
+            }
+        }
+    }
+}
+
 /// A CertificateChain represents a chain of X.509 certificates.
+///
+/// Certificate chains are internally reference counted and are cheaply cloneable.
+#[derive(Clone)]
 pub struct CertificateChain<'a> {
-    ptr: NonNull<s2n_cert_chain_and_key>,
-    is_owned: bool,
+    ptr: Arc<CertificateChainHandle>,
     _lifetime: PhantomData<&'a s2n_cert_chain_and_key>,
 }
 
 impl CertificateChain<'_> {
```

Comment:
What's the plan on adding mutating APIs to `CertificateChain`? i.e `s2n_cert_chain_and_key_set_ocsp_data`

Reply:
Thanks for the callout! I switched to using a builder pattern. Initial mutation is done on a single thread, with the rust type system enforcing a single owner.

The returned CertificateChain is immutable, and therefore can't be modified by _any_ threads.

Reply:
That makes sense, although with fully immutable chain and config updating OCSP staple would require rebuilding the whole config, which is not ideal, but I don't see a better option without introducing thread-safety to s2n-tls methods.

Reply:
If you want to maintain the flexibility to modify things whenever, we could just make the application to handle thread safety.

Instead of making `CertificateChain` internally reference counted, it could be externally reference counted, `Arc<CertificateChain>`. So if customer wanted to mutate it from multiple threads they would need to use an `Arc<Mutex<CertificateChain>>` or something along those lines.

Reply:
But the problem is that we'd have to give non-mutex version to s2n_config internally, at which point you can't safely mutate it again even if there is mutex on rust representation. The only way this will work is if C implementation has locks and is thread-safe.

Reply:
Ya, I was thinking about this further and I don't think there is any safe way to let you mutate an OCSP on an already created certificate. Because there isn't any way for us to guarantee that e.g. s2n-tls isn't currently reading that OCSP data.